storage/redis_adapter.go

```diff
@@ -24,35 +82,61 @@ func NewRedisStorageAdapter(bufferSize int) (*redisAdapter, error) {
 	if err != nil {
 		return nil, err
 	}
-	return &redisAdapter{
+	rsa := &redisAdapter{
 		bufferSize: bufferSize,
 		redisClient: r.NewClient(&r.Options{
 			Addr:     fmt.Sprintf("%s:%d", cfg.RedisHost, cfg.RedisPort),
 			Password: cfg.RedisPassword, // "" == no password
 			DB:       int64(cfg.RedisDB),
 		}),
-	}, nil
+		messageChannel: make(chan *message),
+		stopCh:         make(chan bool),
+	}
+	return rsa, nil
 }
 
-// Write adds a log message to to an app-specific list in redis using ring-buffer-like semantics
-func (a *redisAdapter) Write(app string, message string) error {
-	// Note: Deliberately NOT using MULTI / transactions here since in this implementation of the
-	// redis client, MULTI is not safe for concurrent use by multiple goroutines. It's been advised
-	// by the authors of the gopkg.in/redis.v3 package to just use pipelining when possible...
-	// and here that is technically possible. In the WORST case scenario, not having transactions
-	// means we may momentarily have more than the desired number of log entries in the list /
-	// buffer, but an LTRIM will eventually correct that, bringing the list / buffer back down to
-	// its desired max size.
-	pipeline := a.redisClient.Pipeline()
-	if err := pipeline.RPush(app, message).Err(); err != nil {
-		return err
-	}
-	if err := pipeline.LTrim(app, int64(-1*a.bufferSize), -1).Err(); err != nil {
-		return err
-	}
-	if _, err := pipeline.Exec(); err != nil {
-		return err
+// Start the storage adapter. Invocations of this function are not concurrency safe and multiple
+// serialized invocations have no effect.
+func (a *redisAdapter) Start() {
+	if !a.started {
+		a.started = true
+		errCh := make(chan error)
+		iStopCh := make(chan bool)
+		mp := newMessagePipeliner(a.bufferSize, a.redisClient, errCh)
+		go func() {
+			for {
+				select {
+				case err := <-errCh:
+					log.Println(err)
+				case <-a.stopCh:
+					iStopCh <- true
+					return
+				}
+			}
+		}()
+		go func() {
+			for {
+				select {
+				case message := <-a.messageChannel:
+					mp.addMessage(message)
+					if mp.messageCount == 50 {
+						mp.execPipeline()
+						mp = newMessagePipeliner(a.bufferSize, a.redisClient, errCh)
+					}
+				case <-mp.timeoutTicker.C:
+					mp.execPipeline()
+					mp = newMessagePipeliner(a.bufferSize, a.redisClient, errCh)
+				case <-iStopCh:
```

Comment:
if you `close(a.stopCh)` [below](https://github.com/deis/logger/pull/100/files#r71048651), you'll be able to remove `iStopCh` entirely